pkg/cmd/cli/cmd/set/triggers.go

```diff
@@ -0,0 +1,744 @@
+				From:      o.FromImage,
+				Namespace: o.FromImageNamespace,
+				Auto:      !o.Manual,
+				Names:     names,
+			})
+		}
+	}
+	if o.FromWebHook != nil {
+		triggers.WebHooks = []string{app.GenerateSecret(20)}
+	}
+	if o.FromGitHub != nil {
+		triggers.GitHubWebHooks = []string{app.GenerateSecret(20)}
+	}
+}
+
+type ImageChangeTrigger struct {
+	// If this trigger is automatically applied
+	Auto bool
+	// An ImageStreamTag name to target
+	From string
+	// The target namespace, normalized if set
+	Namespace string
+	// A list of names this trigger targets
+	Names []string
+}
+
+type TriggerDefinition struct {
+	ConfigChange   bool
+	ImageChange    []ImageChangeTrigger
+	WebHooks       []string
+	GitHubWebHooks []string
+}
+
+func defaultNamespace(namespace, defaultNamespace string) string {
+	if namespace == defaultNamespace {
+		return ""
+	}
+	return namespace
+}
+
+func NewDeploymentConfigTriggers(config *deployapi.DeploymentConfig) *TriggerDefinition {
+	t := &TriggerDefinition{}
+	for _, trigger := range config.Spec.Triggers {
+		switch trigger.Type {
+		case deployapi.DeploymentTriggerOnConfigChange:
+			t.ConfigChange = true
+		case deployapi.DeploymentTriggerOnImageChange:
+			t.ImageChange = append(t.ImageChange, ImageChangeTrigger{
+				Auto:      trigger.ImageChangeParams.Automatic,
+				Names:     trigger.ImageChangeParams.ContainerNames,
+				From:      trigger.ImageChangeParams.From.Name,
+				Namespace: defaultNamespace(trigger.ImageChangeParams.From.Namespace, config.Namespace),
+			})
+		}
+	}
+	return t
+}
+
+func NewBuildConfigTriggers(config *buildapi.BuildConfig) *TriggerDefinition {
+	t := &TriggerDefinition{}
+	setStrategy := false
+	for _, trigger := range config.Spec.Triggers {
+		switch trigger.Type {
+		case buildapi.ConfigChangeBuildTriggerType:
+			t.ConfigChange = true
+		case buildapi.GenericWebHookBuildTriggerType:
+			t.WebHooks = append(t.WebHooks, trigger.GenericWebHook.Secret)
+		case buildapi.GitHubWebHookBuildTriggerType:
+			t.GitHubWebHooks = append(t.GitHubWebHooks, trigger.GitHubWebHook.Secret)
+		case buildapi.ImageChangeBuildTriggerType:
+			if trigger.ImageChange.From == nil {
+				if strategyTrigger := strategyTrigger(config); strategyTrigger != nil {
+					setStrategy = true
+					strategyTrigger.Auto = true
+					t.ImageChange = append(t.ImageChange, *strategyTrigger)
+				}
+				continue
+			}
+			// normalize the trigger
+			trigger.ImageChange.From.Namespace = defaultNamespace(trigger.ImageChange.From.Namespace, config.Namespace)
+			t.ImageChange = append(t.ImageChange, ImageChangeTrigger{
+				Auto:      true,
+				From:      trigger.ImageChange.From.Name,
+				Namespace: trigger.ImageChange.From.Namespace,
+			})
+		}
+	}
+	if !setStrategy {
+		if strategyTrigger := strategyTrigger(config); strategyTrigger != nil {
+			t.ImageChange = append(t.ImageChange, *strategyTrigger)
+		}
+	}
+	return t
+}
+
+func (t *TriggerDefinition) Apply(obj runtime.Object) error {
```

Comment:
this seems like it could use some direct unit testing.  i'm especially worried about the strategyTrigger matching logic either being broken now, or getting broken in the future.